kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableFeatures.java

```diff
@@ -49,14 +53,74 @@ public static void validateReadSupportedTable(Protocol protocol, Metadata metada
                         case "vacuumProtocolCheck":
                             break;
                         default:
-                            throw new UnsupportedOperationException(
-                                    "Unsupported table feature: " + readerFeature);
+                            throw DeltaErrors.unsupportedReadFeature(3, readerFeature);
+                    }
+                }
+                break;
+            default:
+                throw unsupportedReaderProtocol(protocol.getMinReaderVersion());
+        }
+    }
+
+    /**
+     * Utility method to validate whether the given table is supported for writing from Kernel.
+     * Currently, the support is as follows:
+     * <ul>
+     *     <li>protocol writer version 1.</li>
+     *     <li>protocol writer version 2 only with appendOnly feature enabled.</li>
+     *     <li>protocol writer version 7 with "appendOnly" feature enabled.</li>
+     * </ul>
+     *
+     * @param protocol    Table protocol
+     * @param metadata    Table metadata
+     * @param tableSchema Table schema
+     */
+    public static void validateWriteSupportedTable(
+            Protocol protocol,
+            Metadata metadata,
+            StructType tableSchema) {
+        int minWriterVersion = protocol.getMinWriterVersion();
+        switch (minWriterVersion) {
+            case 1:
+                break;
+            case 2:
+                // Append-only and column invariants are the writer features added in version 2
+                // Append-only is supported, but not the invariants
+                validateNoInvariants(minWriterVersion, tableSchema);
+                break;
+            case 3:
+                // Check constraints are added in version 3
+                throw unsupportedWriterProtocol(minWriterVersion);
+            case 4:
+                // CDF and generated columns are writer features added in version 4
+                throw unsupportedWriterProtocol(minWriterVersion);
+            case 5:
+                // Column mapping is the only one writer feature added in version 5
+                throw unsupportedWriterProtocol(minWriterVersion);
+            case 6:
+                // Identity is the only one writer feature added in version 6
+                throw unsupportedWriterProtocol(minWriterVersion);
+            case 7:
+                for (String writerFeature : protocol.getWriterFeatures()) {
+                    switch (writerFeature) {
+                        // Only supported writer features as of today in Kernel
+                        case "appendOnly":
```

Comment:
can we just have a static Set of writer features that delta-kernel supports?

Seems like that will be a bit more future-proof and easier to add to?

Reply:
It may not be that simple. For example, if the feature is `columnMapping`, we need to look at metadata to make sure it is one of the supported column mapping features. Similarly, for generatedColumn, we need to check if the expressions in the generated column expression are supported. Most feature will have some additional checks.